src/diffusers/pipelines/stable_diffusion/pipeline_stable_unclip.py

```diff
@@ -0,0 +1,814 @@
+        >>> from diffusers import StableUnCLIPPipeline
+
+        >>> pipe = StableUnCLIPPipeline.from_pretrained(
+        ...     "fusing/stable-unclip-2-1-l", torch_dtype=torch.float16
+        ... )  # TODO update model path
+        >>> pipe = pipe.to("cuda")
+
+        >>> prompt = "a photo of an astronaut riding a horse on mars"
+        >>> images = pipe(prompt).images
+        >>> images[0].save("astronaut_horse.png")
+        ```
+"""
+
+
+class StableUnCLIPPipeline(DiffusionPipeline):
+    """
+    Pipeline for text-to-image generation using stable unCLIP.
+
+    This model inherits from [`DiffusionPipeline`]. Check the superclass documentation for the generic methods the
+    library implements for all the pipelines (such as downloading or saving, running on a particular device, etc.)
+
+    Args:
+        prior_tokenizer ([`CLIPTokenizer`]):
+            Tokenizer of class
+            [CLIPTokenizer](https://huggingface.co/docs/transformers/v4.21.0/en/model_doc/clip#transformers.CLIPTokenizer).
+        prior_text_encoder ([`CLIPTextModelWithProjection`]):
+            Frozen text-encoder.
+        prior ([`PriorTransformer`]):
+            The canonincal unCLIP prior to approximate the image embedding from the text embedding.
+        prior_scheduler ([`DDPMScheduler`]):
+            Scheduler used in the prior denoising process.
+        noise_augmentor ([`NoiseAugmentor`]):
+            Layer for adding noise to the predicted image embeddings. The amount of noise to add is determined by
+            `noise_level` in `StableUnCLIPPipeline.__call__`. See `NoiseAugmentor` for more details on how noise is
+            added.
+        tokenizer (`CLIPTokenizer`):
+            Tokenizer of class
+            [CLIPTokenizer](https://huggingface.co/docs/transformers/v4.21.0/en/model_doc/clip#transformers.CLIPTokenizer).
+        text_encoder ([`CLIPTextModel`]):
+            Frozen text-encoder.
+        unet ([`UNet2DConditionModel`]): Conditional U-Net architecture to denoise the encoded image latents.
+        scheduler ([`DDIMScheduler`]):
+            A scheduler to be used in combination with `unet` to denoise the encoded image latents.
+        vae ([`AutoencoderKL`]):
+            Variational Auto-Encoder (VAE) Model to encode and decode images to and from latent representations.
+    """
+
+    # prior components
+    prior_tokenizer: CLIPTokenizer
+    prior_text_encoder: CLIPTextModelWithProjection
+    prior: PriorTransformer
+    prior_scheduler: DDPMScheduler
+
+    # regular denoising components
+    noise_augmentor: NoiseAugmentor
+    tokenizer: CLIPTokenizer
+    text_encoder: CLIPTextModel
+    unet: UNet2DConditionModel
+    scheduler: DDIMScheduler
+
+    vae: AutoencoderKL
+
+    def __init__(
+        self,
+        # prior components
+        prior_tokenizer: CLIPTokenizer,
+        prior_text_encoder: CLIPTextModelWithProjection,
+        prior: PriorTransformer,
+        prior_scheduler: DDPMScheduler,
```

Comment:
Do you think it could also work with other schedulers or just DDPM for now?

Reply:
Ah yes, it should work with other schedulers, these were just the scheduler types used in the stability source. 

In order for schedulers to be compatible, do they have to support the same beta schedule and the same prediction type? Or is it just the same prediction type and we can swap out different beta schedules after the network is trained (I would be a bit surprised)

Reply:
Yes they have to support the same beta schedule, but it should probably work out of the box give nthat the scheduler config is pretty easy: https://huggingface.co/fusing/stable-unclip-2-1-l/blob/main/prior_scheduler/scheduler_config.json

Reply:
ok cool, will add

Reply:
Ok, changed the types of the schedulers to KarrasDiffusionSchedulers here https://github.com/huggingface/diffusers/pull/2297/commits/c872cbc65812d89babf448d8b2cde2ecdc36708b 

2 related PRs

https://github.com/huggingface/diffusers/pull/2346 fixes the type of KarrasDiffusionSchedulers
https://github.com/huggingface/diffusers/pull/2347 adds the glide noising schedule to the schedulers it's missing from

The prior doesn't work with all KarrasDiffusionSchedulers as some don't give good outputs with it using the sample prediction type. But all KarrasDiffusionSchedulers work with the main denoising process in the pipeline

Reply:
We can merge this PR without merging the other PR's they're just related features + clean up